File: 8.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

type Op struct {
	instruction string
	value       int
}

func run(program []Op) (acc int, e error) {
	var p int
	visited := make(map[int]int)
	for p < len(program) {
		if _, visited := visited[p]; visited {
			return acc, errors.New("error")
		}
		visited[p] = p

		switch op := program[p]; op.instruction {
		case "acc":
			acc += op.value
			p++
		case "jmp":
			p += op.value
		default:
			p++
		}
	}

	return acc, e
}

func main() {
	bytes, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	lines := strings.Split(strings.TrimSpace(string(bytes)), "\n")

	program := make([]Op, len(lines))
	for i, l := range lines {
		split := strings.Fields(l)
		value, _ := strconv.Atoi(split[1])
		program[i] = Op{split[0], value}
	}

	fmt.Println(run(program))

	swap := map[string]string{"jmp": "nop", "nop": "jmp"}
	for i := 0; i < len(program); i++ {
		op := program[i]
		if swapOp, ok := swap[op.instruction]; ok {
			program[i].instruction = swapOp
		} else {
			continue
		}

		if acc, e := run(program); e == nil {
			fmt.Println(acc, "error in row ", i+1)
			break
		}

		program[i].instruction = op.instruction
	}
}

File: 8.py
#!/usr/bin/env python3

import sys


def run(program):
    acc = 0
    p = 0
    visited = set()
    while p < len(program):
        if p in visited:
            return acc, False

        visited.add(p)
        instruction, n = program[p]
        if instruction == "acc":
            acc += n
            p += 1
        elif instruction == "jmp":
            p += n
        else:
            p += 1

    return acc, True


def swap(x):
    if x == "jmp":
        return "nop"
    elif x == "nop":
        return "jmp"
    else:
        raise ValueError("Illegal instruction", x)


if __name__ == "__main__":
    with open(sys.argv[1]) as f:
        lines = [x.strip() for x in f.readlines()]

    program = [(x, int(n)) for x, n in [l.split() for l in lines]]

    print(run(program))

    for i in range(len(program)):
        instruction, n = program[i]
        try:
            program[i] = (swap(instruction), n)
        except ValueError:
            continue

        acc, ok = run(program)
        if ok:
            print(acc, "error in row", i + 1)
            break

        program[i] = (instruction, n)
